src/filters/cosmetic.rs

```diff
@@ -0,0 +1,1493 @@
+                return Err(CosmeticFilterError::InvalidCssSelector);
+            } else if let Some(ref style) = style {
+                if !is_valid_css_style(style) {
+                    return Err(CosmeticFilterError::InvalidCssStyle);
+                } else if sharp_index == 0 {
+                    return Err(CosmeticFilterError::GenericStyle);
+                }
+            }
+
+            if (not_entities.is_some() || not_hostnames.is_some()) && mask.contains(CosmeticFilterMask::UNHIDE) {
+                return Err(CosmeticFilterError::DoubleNegation);
+            }
+
+            if !selector.is_ascii() {
+                mask |= CosmeticFilterMask::IS_UNICODE;
+            }
+
+            let key = if !mask.contains(CosmeticFilterMask::SCRIPT_INJECT) {
+                if selector.starts_with('.') {
+                    let key = key_from_selector(selector)?;
+                    mask |= CosmeticFilterMask::IS_CLASS_SELECTOR;
+                    if key == selector {
+                        mask |= CosmeticFilterMask::IS_SIMPLE;
+                    }
+                    Some(String::from(&key[1..]))
+                } else if selector.starts_with('#') {
+                    let key = key_from_selector(selector)?;
+                    mask |= CosmeticFilterMask::IS_ID_SELECTOR;
+                    if key == selector {
+                        mask |= CosmeticFilterMask::IS_SIMPLE;
+                    }
+                    Some(String::from(&key[1..]))
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
+            Ok(CosmeticFilter {
+                entities,
+                hostnames,
+                mask,
+                not_entities,
+                not_hostnames,
+                raw_line: if debug {
+                    Some(String::from(line))
+                } else {
+                    None
+                },
+                selector: String::from(selector),
+                key,
+                style,
+            })
+        } else {
+            Err(CosmeticFilterError::MissingSharp)
+        }
+    }
+
+    /// Any cosmetic filter rule that specifies (possibly negated) hostnames or entities has a
+    /// hostname constraint.
+    pub fn has_hostname_constraint(&self) -> bool {
+        self.hostnames.is_some() ||
+            self.entities.is_some() ||
+            self.not_entities.is_some() ||
+            self.not_hostnames.is_some()
+    }
+
+    /// Check whether this rule applies to content from the hostname and domain corresponding to
+    /// the provided hash lists.
+    ///
+    /// See the `matches_str` test helper function for an example of how to convert hostnames and
+    /// domains into the appropriate hash lists.
+    pub fn matches(&self, request_entities: &[Hash], request_hostnames: &[Hash]) -> bool {
```

Comment:
method makes sense, but not seeing where it gets used besides tests

Reply:
This was from before I started storing rules keyed by hostname.

It might still be useful for some kind of logging or dashboard? Should I just remove it altogether?

Reply:
let's remove it, if we want it again we can probably do better